### PR TITLE
[Clang] Implement P2718R0 "Lifetime extension in range-based for loops"

### PR DESCRIPTION
**Pull request by yronglin**
Implement P2718R0 "Lifetime extension in range-based for loops" (https://wg21.link/P2718R0)

Differential Revision: https://reviews.llvm.org/D153701

### CODE REVIEW

**Comment by yronglin** [comment]
ping~

File: clang/lib/Parse/ParseStmt.cpp
Comment (AaronBallman) [inline clang/lib/Parse/ParseStmt.cpp:2297]:
```suggestion
  } else if (ForEach) {
    // Similarly, we need to do the semantic analysis for a for-range
    // statement immediately in order to close over temporaries correctly.
    ForEachStmt = Actions.ActOnObjCForCollectionStmt(ForLoc,
```
Might as well move the comment to where it belongs.

File: clang/lib/Sema/SemaExpr.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaExpr.cpp:6299]:
```suggestion
      // Pass down lifetime extending flag, and collect temporaries in
```

File: clang/lib/Sema/SemaExpr.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaExpr.cpp:6319]:
```suggestion
      // Collect MaterializeTemporaryExprs in the rewritten CXXDefaultArgExpr.
```

File: clang/lib/Sema/SemaExprCXX.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaExprCXX.cpp:8237]: Can you explain why the check for C++17?

File: clang/lib/Sema/SemaStmt.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaStmt.cpp:2543]: Please spell out the type instead of using `auto`. Also, should this be a const ref to avoid a copy?

File: clang/lib/Sema/SemaStmt.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaStmt.cpp:2792]: Spurious whitespace change can be removed.

**Review by AaronBallman — COMMENTED** [review]
Thank you for this! I think it looks pretty close, but I'd love to get more eyes on this given the importance of the feature. CC @zygoloid @rjmccall @erichkeane

File: clang/lib/Parse/ParseStmt.cpp
Comment (yronglin) [reply clang/lib/Parse/ParseStmt.cpp]: Agree, done.

File: clang/lib/Sema/SemaExpr.cpp
Comment (yronglin) [reply clang/lib/Sema/SemaExpr.cpp]: fixed

File: clang/lib/Sema/SemaExpr.cpp
Comment (yronglin) [reply clang/lib/Sema/SemaExpr.cpp]: fixed

File: clang/lib/Sema/SemaStmt.cpp
Comment (yronglin) [reply clang/lib/Sema/SemaStmt.cpp]: I use`InitializedEntity ` instead of auto, `InitializedEntity::InitializeVariable` returns a rvalue, so I haven't add a const ref.

File: clang/lib/Sema/SemaStmt.cpp
Comment (yronglin) [reply clang/lib/Sema/SemaStmt.cpp]: removed

**Comment by yronglin** [comment]
Add `ReleaseNotes`.

File: clang/lib/Sema/SemaStmt.cpp
Comment (AaronBallman) [reply clang/lib/Sema/SemaStmt.cpp]: Ah `InitializedEntity` doesn't have a copy constructor so the move operation will be cheap. Thanks!

File: clang/lib/Sema/SemaExprCXX.cpp
Comment (AaronBallman) [reply clang/lib/Sema/SemaExprCXX.cpp:8237]: Ah okay, so this is still related to the lifetime extension for range-based for loops -- we needed this bit for C++17 conformance which we previously didn't have to do?

(Mostly trying to reason about whether this is a separate bug fix that should be landed in a separate patch or whether this is related to the correctness for P2718R0.)

Comment (yronglin) [reply clang/lib/Sema/SemaExprCXX.cpp:8237]: Yeah, I think so. we previously didn't materialize temporary object may should also be in compliance with the standard words "the materialization of a temporary object is generally delayed as long as possible in order to avoid creating unnecessary temporary object".

**Comment by yronglin** [comment]
Rebase to fix CI issue

**Comment by cor3ntin** [comment]
@zygoloid Can you take another crack at this? If you have time to confirm the design maybe we could try to land it in Clang 18. Otherwise the plan is to be conservative. Thanks!

File: clang/include/clang/Sema/Sema.h
Comment (zygoloid) [inline clang/include/clang/Sema/Sema.h:9985]: Is there a reason that this is capitalized and the function above is not? (Genuine question, I've lost track of LLVM's preferred capitalization rule for functions, but this locally looks inconsistent.)

File: clang/include/clang/Sema/Sema.h
Comment (zygoloid) [inline clang/include/clang/Sema/Sema.h:1368]: I think we should be clear here that this is only for unusual lifetime-extension contexts, and not just (for example) in an initializer of a reference.
```suggestion
    /// Whether we are currently in a context in which all temporaries must be
    /// lifetime-extended, even if they're not bound to a reference (for example,
    /// in a for-range initializer).
    bool IsInLifetimeExtendingContext = false;
```

**Review by zygoloid — COMMENTED** [review]
This looks good to me.

File: clang/lib/Parse/ParseDecl.cpp
Comment (shafik) [inline clang/lib/Parse/ParseDecl.cpp:2337]: Why not move this into the `if (getLangOpts().CPlusPlus23)` above?

File: clang/test/CXX/special/class.temporary/p6.cpp
Comment (shafik) [inline clang/test/CXX/special/class.temporary/p6.cpp:243]: Is it possible to also test the case where it does not clean up like in the proposal:

```cpp
for (auto e : f2(g())) {}  // undefined behavior
```

File: clang/include/clang/Sema/Sema.h
Comment (cor3ntin) [inline clang/include/clang/Sema/Sema.h:1342]:
```suggestion
    /// MaterializeTemporaryExprs in for-range-init expressions which need to
```

File: clang/lib/Parse/ParseDecl.cpp
Comment (cor3ntin) [reply clang/lib/Parse/ParseDecl.cpp:2337]: `ForRangeLifetimeExtendTemps` is going to be filled by `ParseExpression`/ I wonder if we should do that unconditionally though.
and replace the if with `assert(getLangOpts().CPlusPlus23 || Actions.ExprEvalContexts.back().ForRangeLifetimeExtendTemps.empty())`

File: clang/lib/Sema/SemaExpr.cpp
Comment (cor3ntin) [inline clang/lib/Sema/SemaExpr.cpp:6306]: We probably should do the same thing for `BuildCXXDefaultInitExpr`, right?

```cpp

struct my_range {
  // ...
  foo bar = temp();
};

for(elem : my_range {}) {}
```

File: clang/include/clang/Sema/Sema.h
Comment (yronglin) [reply clang/include/clang/Sema/Sema.h:9985]: LLVM prefer `shouldMaterializePRValueInDiscardedExpression` https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly , I'll rename it.

File: clang/lib/Parse/ParseDecl.cpp
Comment (yronglin) [reply clang/lib/Parse/ParseDecl.cpp:2337]: agree

File: clang/test/CXX/special/class.temporary/p6.cpp
Comment (yronglin) [reply clang/test/CXX/special/class.temporary/p6.cpp:243]: agree, I have compiled this case, Currently, clang will not extending lifetime of temporaries in this case.

File: clang/lib/Sema/SemaExpr.cpp
Comment (yronglin) [reply clang/lib/Sema/SemaExpr.cpp:6306]: Good catch! Yeah, we need do the same thing for `BuildCXXDefaultInitExpr`, I'm working on it.

**Comment by cor3ntin** [comment]
@yronglin Do you know when you will be able to get to this? thanks!

**Comment by yronglin** [comment]
> @yronglin Do you know when you will be able to get to this? thanks!

Sorry for the lack of new actions for a long time. I have recently fixed some problems in the dependent context. I will try to complete the DefaultInit related functions and tests this weekend.

File: clang/include/clang/Sema/Sema.h
Comment (yronglin) [inline clang/include/clang/Sema/Sema.h:1382]: I think this variable and related function `needMaterializePRValueInDiscardedExpression()` should be renamed to  `NeedMaterializePRValue` and `needMaterializePRValue()` , then it can use in other cases.

In [class.temporary P2][http://eel.is/c++draft/class.temporary#2]

The materialization of a temporary object is generally delayed as long as possible in order to avoid creating unnecessary temporary objects[.](http://eel.is/c++draft/class.temporary#2.sentence-1)
[Note [3](http://eel.is/c++draft/class.temporary#note-3): Temporary objects are materialized:
[(2.1)](http://eel.is/c++draft/class.temporary#2.1)
when binding a reference to a prvalue ([[dcl.init.ref]](http://eel.is/c++draft/dcl.init.ref), [[expr.type.conv]](http://eel.is/c++draft/expr.type.conv), [[expr.dynamic.cast]](http://eel.is/c++draft/expr.dynamic.cast), [[expr.static.cast]](http://eel.is/c++draft/expr.static.cast), [[expr.const.cast]](http://eel.is/c++draft/expr.const.cast), [[expr.cast]](http://eel.is/c++draft/expr.cast)),
[(2.2)](http://eel.is/c++draft/class.temporary#2.2)
when performing member access on a class prvalue ([[expr.ref]](http://eel.is/c++draft/expr.ref), [[expr.mptr.oper]](http://eel.is/c++draft/expr.mptr.oper)),
[(2.3)](http://eel.is/c++draft/class.temporary#2.3)
when performing an array-to-pointer conversion or subscripting on an array prvalue ([[conv.array]](http://eel.is/c++draft/conv.array), [[expr.sub]](http://eel.is/c++draft/expr.sub)),
[(2.4)](http://eel.is/c++draft/class.temporary#2.4)
when initializing an object of type std​::​initializer_list<T> from a [braced-init-list](http://eel.is/c++draft/dcl.init.general#nt:braced-init-list) ([[dcl.init.list]](http://eel.is/c++draft/dcl.init.list)),
[(2.5)](http://eel.is/c++draft/class.temporary#2.5)
for certain unevaluated operands ([[expr.typeid]](http://eel.is/c++draft/expr.typeid), [[expr.sizeof]](http://eel.is/c++draft/expr.sizeof)), and
[(2.6)](http://eel.is/c++draft/class.temporary#2.6)
when a prvalue that has type other than cv void appears as a discarded-value expression ([[expr.context]](http://eel.is/c++draft/expr.context))[.](http://eel.is/c++draft/class.temporary#2.sentence-2)
— end note]

**Comment by yronglin** [comment]
Should we also dump the sub-expr which in `CXXDefaultArgExpr`:
for example:
Consider:
```
struct A {
  A();
  ~A();
};

typedef int vec3[3];
vec3 &f(const A &a = A{});

void foo() {
  for (auto e : f()) {}
}
```
Current AST Dump:
```
FunctionDecl 0x126813170 <line:85:1, line:87:1> line:85:6 foo 'void ()'
  `-CompoundStmt 0x126814720 <col:12, line:87:1>
    `-CXXForRangeStmt 0x1268146b0 <line:86:3, col:23>
      |-<<<NULL>>>
      |-DeclStmt 0x1268137b8 <col:17>
      | `-VarDecl 0x126813548 <col:17, col:19> col:17 implicit used __range1 'vec3 &' cinit
      |   `-ExprWithCleanups 0x126813730 <col:17, col:19> 'vec3':'int[3]' lvalue
      |     `-CallExpr 0x126813318 <col:17, col:19> 'vec3':'int[3]' lvalue
      |       |-ImplicitCastExpr 0x126813300 <col:17> 'vec3 &(*)(const A &)' <FunctionToPointerDecay>
      |       | `-DeclRefExpr 0x126813278 <col:17> 'vec3 &(const A &)' lvalue Function 0x126813030 'f' 'vec3 &(const A &)'
      |       `-CXXDefaultArgExpr 0x1268133f0 <<invalid sloc>> 'const A' lvalue
      |-DeclStmt 0x126814448 <col:15>
      | `-VarDecl 0x126813880 <col:15> col:15 implicit used __begin1 'int *' cinit
 .......
```
We can't see directly on the AST whether the lifetime is extended or not. The sub-expr which in `CXXDefaultArgExpr` is:
```
ExprWithCleanups 0x1398c55d8 'const A':'const struct A' lvalue
`-MaterializeTemporaryExpr 0x1398c55c0 'const A':'const struct A' lvalue extended by Var 0x1398c5748 '__range1' 'vec3 &'
  `-ImplicitCastExpr 0x1398c55a8 'const A':'const struct A' <NoOp>
    `-CXXBindTemporaryExpr 0x1398c5588 'A':'struct A' (CXXTemporary 0x1398c5588)
      `-CXXTemporaryObjectExpr 0x1398c5550 'A':'struct A' 'void (void)'
```

**Comment by cor3ntin** [comment]
> Should we also dump the sub-expr which in `CXXDefaultArgExpr`: for example: Consider:
> 
> ```
> struct A {
>   A();
>   ~A();
> };
> 
> typedef int vec3[3];
> vec3 &f(const A &a = A{});
> 
> void foo() {
>   for (auto e : f()) {}
> }
> ```
> 
> Current AST Dump:
> 
> ```
> FunctionDecl 0x126813170 <line:85:1, line:87:1> line:85:6 foo 'void ()'
>   `-CompoundStmt 0x126814720 <col:12, line:87:1>
>     `-CXXForRangeStmt 0x1268146b0 <line:86:3, col:23>
>       |-<<<NULL>>>
>       |-DeclStmt 0x1268137b8 <col:17>
>       | `-VarDecl 0x126813548 <col:17, col:19> col:17 implicit used __range1 'vec3 &' cinit
>       |   `-ExprWithCleanups 0x126813730 <col:17, col:19> 'vec3':'int[3]' lvalue
>       |     `-CallExpr 0x126813318 <col:17, col:19> 'vec3':'int[3]' lvalue
>       |       |-ImplicitCastExpr 0x126813300 <col:17> 'vec3 &(*)(const A &)' <FunctionToPointerDecay>
>       |       | `-DeclRefExpr 0x126813278 <col:17> 'vec3 &(const A &)' lvalue Function 0x126813030 'f' 'vec3 &(const A &)'
>       |       `-CXXDefaultArgExpr 0x1268133f0 <<invalid sloc>> 'const A' lvalue
>       |-DeclStmt 0x126814448 <col:15>
>       | `-VarDecl 0x126813880 <col:15> col:15 implicit used __begin1 'int *' cinit
>  .......
> ```
> 
> We can't see directly on the AST whether the lifetime is extended or not. The sub-expr which in `CXXDefaultArgExpr` is:
> 
> ```
> ExprWithCleanups 0x1398c55d8 'const A':'const struct A' lvalue
> `-MaterializeTemporaryExpr 0x1398c55c0 'const A':'const struct A' lvalue extended by Var 0x1398c5748 '__range1' 'vec3 &'
>   `-ImplicitCastExpr 0x1398c55a8 'const A':'const struct A' <NoOp>
>     `-CXXBindTemporaryExpr 0x1398c5588 'A':'struct A' (CXXTemporary 0x1398c5588)
>       `-CXXTemporaryObjectExpr 0x1398c5550 'A':'struct A' 'void (void)'
> ```

I think this makes sense, but it can be a separate patch, unless it helps for testing.


FYI, we prefer merge commits instead of rebase + force push because Github reviews tend to loose context when doing force pushes / rebases in PRs

File: clang/lib/Sema/SemaExpr.cpp
Comment (cor3ntin) [inline clang/lib/Sema/SemaExpr.cpp:6395]: Where are the tests for BuildCXXDefaultInitExpr?

Comment (yronglin) [reply clang/lib/Sema/SemaExpr.cpp:6395]: Thanks for your review! I still have some questions about member init.
For this case(https://godbolt.org/z/jPKe8M6xE), should we materialize the temporaries in class Z::x's default init,  but unless I'm mistaken?

**Comment by yronglin** [comment]
For this case, should

> > Should we also dump the sub-expr which in `CXXDefaultArgExpr`: for example: Consider:
> 
> I think this makes sense, but it can be a separate patch, unless it helps for testing.

Cool! Let's do it in a separate patch.

> FYI, we prefer merge commits instead of rebase + force push because Github reviews tend to loose context when doing force pushes / rebases in PRs

Thanks for the reminder, I may not have converted from the Phabricator habit yet. I tried to rebase to fix a libcxx ci issue(gdb_pretty_printer_test)

File: clang/lib/Sema/SemaExpr.cpp
Comment (yronglin) [reply clang/lib/Sema/SemaExpr.cpp:6395]: @cor3ntin I tried this code https://godbolt.org/z/n77v9rMTz , and the temporaries which in `CXXDefaultInitExpr`  has been lifetime extended in AST. But the LLVM IR unexpected, and the lifetime of the temporaries was not extended. Could you please guide me as to why this is and whether I have made some mistake?

Comment (cor3ntin) [reply clang/lib/Sema/SemaExpr.cpp:6395]: I think part of the problem might be that we should only extend in the aggregate case.
ie, when we call `BuildCXXDefaultInit` from `FillInEmptyInitForField`.

In the constructor case, the init lifetime is bound to that of the member initializer expression.

Comment (yronglin) [reply clang/lib/Sema/SemaExpr.cpp:6395]: Thanks for your help! I have a case use aggregate initialization (https://godbolt.org/z/vn8Y11ja8). I think neither the brace-or-equals initializer nor default member initializer should not be materialized and lifetime-extended in this code snippet, seems they don't satisfy any of [class.temporary P2](http://eel.is/c++draft/class.temporary#2) . @hubert-reinterpretcast Could you please give me an example where `CXXDefaultInitExpr` has to be lifetime-extended? I saw you comment in D153701 that we need to do lifetime-extend for `CXXDefaultInitExpr`. I'm sorry that I'm unfamiliar with that part of clang and I have no clue how to write tests for CXXDefaultInitExpr.

Comment (cor3ntin) [reply clang/lib/Sema/SemaExpr.cpp:6395]: I think you are right. 
I'm sorry, that whole thing is a bit confusing to me.

But the initializer is used in the construction of a subobject, which is different from a default argument which is initialized on call site... so I think the conclusion is that we should do nothing for default inits, I think

Comment (yronglin) [reply clang/lib/Sema/SemaExpr.cpp:6395]: Thank you for your consideration.  I think so too, whether I have made some mistake, I will try to ask the author of P2718R0

**Comment by cor3ntin** [comment]
Here is what I'd like to do @yronglin
Revert the changes to BuildCXXDefaultInit.
And land it in main.
Then we'll have a few weeks to decide if we backport it to 18. Dies that sound good to you?

File: clang/www/cxx_status.html
Comment (cor3ntin) [inline clang/www/cxx_status.html:432]:
```suggestion
      <td class="none" align="center">Clang 19</td>
```

Comment (yronglin) [reply clang/www/cxx_status.html]: done

**Comment by yronglin** [comment]
> Here is what I'd like to do @yronglin Revert the changes to BuildCXXDefaultInit. And land it in main. Then we'll have a few weeks to decide if we backport it to 18. Dies that sound good to you?

Yeah, it's looks good to me, I have reverted changes for `CXXDefaultInitExpr`.

**Review by cor3ntin — APPROVED** [review]
LGTM. 
Do you need me to merge that for you?

**Comment by yronglin** [comment]
> LGTM. Do you need me to merge that for you?

Thanks a lot for your review and help! I have commit access, once the CI green, I'll land this patch.

File: clang/lib/Sema/SemaExpr.cpp
Comment (hubert-reinterpretcast) [reply clang/lib/Sema/SemaExpr.cpp]: > @hubert-reinterpretcast Could you please give me an example where `CXXDefaultInitExpr` has to be lifetime-extended?

https://github.com/llvm/llvm-project/issues/85613

Comment (yronglin) [reply clang/lib/Sema/SemaExpr.cpp]: Many thanks for your information.

Comment (yronglin) [reply clang/lib/Sema/SemaExpr.cpp]: I try to extend lifetime of the `MaterializedTemporaryExpr`s in `CXXDefaultInitExpr`, but looking from the generated IR, `MaterializedTemporaryExpr ::setExtendingDecl` does not take effect. I think some modifications may be needed in CodeGen.

**Pull request merged**
